File: src/core/interprocedural-dse.ts
import type { AppliedOptimization, ProgramAst, StatementAst } from "./types.ts";
import { buildRuleCfg, exprIsCallFree, programStateFields, type RuleCfg } from "./rule-cfg.ts";

// Interprocedural, flow-sensitive dead-store elimination on the rule call
// graph. A pure assignment `field = expr` is removed when, on every path that
// leaves it, the field is reassigned before it is ever observed (shown, read in
// a condition, read into another expression, output, or read back as input).
// This is exactly register liveness over the whole-program CFG; the value is
// dead when its register is not live-out.

function computeLiveOut(cfg: RuleCfg, allVars: ReadonlySet<string>): Set<string>[] {
  const n = cfg.nodes.length;
  const liveIn: Set<string>[] = Array.from({ length: n }, () => new Set<string>());
  const liveOut: Set<string>[] = Array.from({ length: n }, () => new Set<string>());
  let changed = true;
  let rounds = 0;
  while (changed && rounds < 1000) {
    changed = false;
    rounds += 1;
    for (let i = n - 1; i >= 0; i -= 1) {
      const node = cfg.nodes[i]!;
      const newOut = new Set<string>();
      for (const successor of node.succ) {
        for (const reg of liveIn[successor]!) newOut.add(reg);
      }
      // A barrier (raw core / egg / unknown call) may read anything, so keep
      // every variable live across it.
      const uses = node.barrier ? allVars : node.uses;
      const defs = node.barrier ? [] : node.defs;
      const newIn = new Set<string>(uses);
      for (const reg of newOut) {
        if (!defs.includes(reg)) newIn.add(reg);
      }
      if (!setsEqual(newIn, liveIn[i]!) || !setsEqual(newOut, liveOut[i]!)) {
        liveIn[i] = newIn;
        liveOut[i] = newOut;
        changed = true;
      }
    }
  }
  return liveOut;
}

function setsEqual(a: ReadonlySet<string>, b: ReadonlySet<string>): boolean {
  if (a.size !== b.size) return false;
  for (const value of a) if (!b.has(value)) return false;
  return true;
}

function collectAllVars(cfg: RuleCfg): Set<string> {
  const vars = new Set<string>();
  for (const node of cfg.nodes) {
    for (const def of node.defs) vars.add(def);
    for (const use of node.uses) vars.add(use);
  }
  return vars;
}

function pruneStatements(statements: StatementAst[], doomed: ReadonlySet<StatementAst>): StatementAst[] {
  return statements.flatMap((statement): StatementAst[] => {
    if (doomed.has(statement)) return [];
    if (statement.kind === "loop") return [{ ...statement, body: pruneStatements(statement.body, doomed) }];
    if (statement.kind === "if") {
      const pruned: Extract<StatementAst, { kind: "if" }> = {
        ...statement,
        thenBody: pruneStatements(statement.thenBody, doomed),
      };
      if (statement.elseBody !== undefined) pruned.elseBody = pruneStatements(statement.elseBody, doomed);
      return [pruned];
    }
    if (statement.kind === "dispatch") {
      return [{
        ...statement,
        cases: statement.cases.map((branch) => ({ ...branch, body: pruneStatements(branch.body, doomed) })),
        ...(statement.defaultBody === undefined ? {} : { defaultBody: pruneStatements(statement.defaultBody, doomed) }),
      }];
    }
    return [statement];
  });
}

export function eliminateInterproceduralDeadStores(
  ast: ProgramAst,
  optimizations: AppliedOptimization[],
): number {
  const fields = programStateFields(ast);
  if (fields.size === 0) return 0;
  const cfg = buildRuleCfg(ast);
  const allVars = collectAllVars(cfg);
  const liveOut = computeLiveOut(cfg, allVars);

  const doomed = new Set<StatementAst>();
  for (const node of cfg.nodes) {
    const assign = node.assign;
    if (assign === undefined) continue;
    if (!fields.has(assign.target)) continue;
    if (!exprIsCallFree(assign.expr)) continue;
    if (liveOut[node.id]!.has(assign.target)) continue;
    doomed.add(assign);
  }
  if (doomed.size === 0) return 0;

  for (const entry of ast.entries) entry.body = pruneStatements(entry.body, doomed);
  for (const proc of ast.procs) proc.body = pruneStatements(proc.body, doomed);

  optimizations.push({
    name: "interprocedural-dead-store",
    detail: `Removed ${doomed.size} store(s) whose value is always overwritten before it can be observed across the rule call graph.`,
  });
  return doomed.size;
}

File: native/src/core/interprocedural_dse.cpp
#include "mkpro/core/interprocedural_dse.hpp"

#include "mkpro/core/parser.hpp"
#include "mkpro/core/rule_cfg.hpp"

#include <algorithm>
#include <optional>
#include <set>
#include <string>
#include <utility>
#include <vector>

namespace mkpro::core {

namespace {

std::optional<Expression> parse_expression_safe(const std::string& text, int line) {
  try {
    return parse_expression(text, line);
  } catch (const std::exception&) {
    return std::nullopt;
  }
}

std::optional<std::string> simple_show_target(const V2Statement& statement) {
  if (statement.kind != "v2_show")
    return std::nullopt;
  if (statement.target.has_value()) {
    const std::optional<Expression> target = parse_expression_safe(*statement.target, statement.line);
    if (target.has_value() && target->kind == "identifier")
      return target->name;
  }
  if (statement.items.has_value() && statement.items->size() == 1U) {
    const DisplayItem& item = statement.items->front();
    if (item.kind != "source")
      return std::nullopt;
    if (item.expr.has_value() && item.expr->kind == "identifier")
      return item.expr->name;
    if (!item.expr.has_value() && !item.name.empty())
      return item.name;
  }
  return std::nullopt;
}

void collect_loop_header_show_targets(const std::vector<V2Statement>& statements,
                                      std::set<std::string>& targets);

void collect_loop_header_show_target(const V2Statement& statement, std::set<std::string>& targets) {
  if (statement.kind == "v2_loop" && statement.body.size() >= 2U) {
    const V2Statement& show = statement.body.at(0);
    const V2Statement& read = statement.body.at(1);
    if (show.kind == "v2_show" && read.kind == "v2_read") {
      if (std::optional<std::string> target = simple_show_target(show))
        targets.insert(*target);
    }
  }
  collect_loop_header_show_targets(statement.body, targets);
  collect_loop_header_show_targets(statement.then_body, targets);
  collect_loop_header_show_targets(statement.else_body, targets);
  for (const V2MatchCase& match_case : statement.cases) {
    if (match_case.action != nullptr)
      collect_loop_header_show_target(*match_case.action, targets);
  }
  if (statement.otherwise != nullptr)
    collect_loop_header_show_target(*statement.otherwise, targets);
}

void collect_loop_header_show_targets(const std::vector<V2Statement>& statements,
                                      std::set<std::string>& targets) {
  for (const V2Statement& statement : statements)
    collect_loop_header_show_target(statement, targets);
}

std::set<std::string> loop_header_show_targets(const V2Program& program) {
  std::set<std::string> targets;
  collect_loop_header_show_targets(program.body, targets);
  for (const V2Rule& rule : program.rules)
    collect_loop_header_show_targets(rule.body, targets);
  return targets;
}

V2Statement statement_from_pruned_vector(std::vector<V2Statement> statements, int line) {
  if (statements.size() == 1)
    return std::move(statements.front());
  return V2Statement{
      .kind = "v2_block",
      .body = std::move(statements),
      .line = line,
  };
}

std::vector<V2Statement> prune_statements(const std::vector<V2Statement>& statements,
                                          const std::set<const V2Statement*>& doomed);

V2Statement prune_statement(const V2Statement& statement,
                            const std::set<const V2Statement*>& doomed) {
  V2Statement result = statement;
  if (statement.kind == "v2_loop" || statement.kind == "v2_while" || statement.kind == "v2_block") {
    result.body = prune_statements(statement.body, doomed);
  } else if (statement.kind == "v2_if") {
    result.then_body = prune_statements(statement.then_body, doomed);
    result.else_body = prune_statements(statement.else_body, doomed);
  } else if (statement.kind == "v2_match") {
    for (V2MatchCase& match_case : result.cases) {
      if (match_case.action == nullptr)
        continue;
      std::vector<V2Statement> pruned = prune_statements({*match_case.action}, doomed);
      match_case.action = std::make_shared<V2Statement>(
          statement_from_pruned_vector(std::move(pruned), match_case.line));
    }
    if (result.otherwise != nullptr) {
      std::vector<V2Statement> pruned = prune_statements({*result.otherwise}, doomed);
      result.otherwise = std::make_shared<V2Statement>(
          statement_from_pruned_vector(std::move(pruned), statement.line));
    }
  }
  return result;
}

std::vector<V2Statement> prune_statements(const std::vector<V2Statement>& statements,
                                          const std::set<const V2Statement*>& doomed) {
  std::vector<V2Statement> result;
  result.reserve(statements.size());
  for (const V2Statement& statement : statements) {
    if (doomed.contains(&statement))
      continue;
    result.push_back(prune_statement(statement, doomed));
  }
  return result;
}

} // namespace

int eliminate_interprocedural_dead_stores(V2Program& program,
                                          std::vector<OptimizationReport>& optimizations) {
  const std::set<std::string> fields = program_state_fields(program);
  if (fields.empty())
    return 0;

  const RuleCfg cfg = build_rule_cfg(program);
  const RuleCfgLiveness liveness = compute_rule_cfg_liveness(cfg);
  const std::set<std::string> loop_show_targets = loop_header_show_targets(program);

  std::set<const V2Statement*> doomed;
  for (const RuleCfgNode& node : cfg.nodes) {
    const V2Statement* assign = node.assign;
    if (assign == nullptr || !assign->target.has_value() || !assign->expr.has_value())
      continue;
    const std::optional<Expression> target = parse_expression_safe(*assign->target, assign->line);
    if (!target.has_value() || target->kind != "identifier")
      continue;
    if (!fields.contains(target->name))
      continue;
    if (loop_show_targets.contains(target->name))
      continue;
    const std::optional<Expression> expr = parse_expression_safe(*assign->expr, assign->line);
    if (!expr.has_value() || !expression_is_call_free(*expr))
      continue;
    if (liveness.live_out.at(static_cast<std::size_t>(node.id)).contains(target->name))
      continue;
    doomed.insert(assign);
  }

  if (doomed.empty())
    return 0;

  program.body = prune_statements(program.body, doomed);
  for (V2Rule& rule : program.rules)
    rule.body = prune_statements(rule.body, doomed);

  optimizations.push_back(OptimizationReport{
      .name = "interprocedural-dead-store",
      .detail = "Removed " + std::to_string(doomed.size()) +
                " store(s) whose value is always overwritten before it can be observed across "
                "the rule call graph.",
  });
  return static_cast<int>(doomed.size());
}

} // namespace mkpro::core
